AppController/djinn.rb

```diff
@@ -4597,49 +4566,44 @@ def place_error_app(version_key, err_msg)
 
   # This function ensures that applications we are not aware of (that is
   # they are not accounted for) will be terminated and, potentially old
-  # sources, will be removed.
+  # sources, will be removed. Must be called under APPS_LOCK.
   def check_stopped_apps
     Djinn.log_debug("Checking applications that have been stopped.")
-    version_list = HelperFunctions.get_loaded_versions
-    begin
-      configured_versions = ZKInterface.get_versions
-    rescue FailedZooKeeperOperationException
-      Djinn.log_warn(
-        'Failed to fetch configured versions when checking stopped apps')
-      return
-    end
-
-    version_list.each { |version_key|
+    removed_versions = []
+    @versions_loaded.each { |version_key|
```

Comment:
Rather than a single loop on `@versions_loaded`, I think the actual resources that are being affected should be checked. For example, something like
```
if my_node.is_shadow?
  syslog_files.each do |version_key, file_location|
    next if version_key in configured_versions
    FileUtils.rm(file_location)
    HelperFunctions.shell("service rsyslog restart")
  end
end
```
I don't think we have any guarantees that the configuration files on disk match `@versions_loaded`. For example, the controller could finish setting up an application and then crash before writing its state. When it comes back up, there could be artifacts on disk that it should clean up even though the restored `@versions_loaded` does not account for them.